### DISCUSSION TITLE
Lua_ls with Termux

### DISCUSSION THREAD

**Discussion by aBaechtold** [mason-org/mason.nvim#1438] [General]
Hi,
I'm using Termux (Android phone, v0.118.0) to run NeoVim (nvim) and was trying to use Mason as a tools extension manager plugin i.e. for a lightweight IDE. I setup the mason_lspconfig plugin along with it and intended to have lua_ls (Sumneko's lua-language-server) automatically installed to support nvim's Lua side of things. However this fails. The error is: `"The current platform is unsupported."`

I'm fairly new if it comes to nvim/vim and Lua, or Mason for that matter, but from investigation the error message seems to be from Mason's core module (see log below). The repository entry for lua_ls defines common targets as released by Sumneko's GitHub project e.g. `linux_arm64_gnu`. There seems to be no match for the Termux setup. `uname` on my Termux reports as `Linux`, `aarch64`, or per GCC (Clang) as `gcc -v; clang version 16.0.6;  Target: aarch64-unknown-linux-android24`.

It's a bit unfortunate that one of the main LSP packages fail to install, but wouldn't consider this a failure of Mason. Termux is probably not considered by many as a target. Hence a discussion rather than an issue.

The question is what options I have to get lua_ls working. I see the following ones, given that I'm able to build it from the sources and that apparently there is even a Termux Pkg (built on APT) package:

- Not using Mason and install the Pkg package.
- Being able to define a local Mason reppsitory to either point to the Pkg package, or to a self-hosted privately build package (e.g on GitHub or local file system).
- Get offical support from Sumneko and update the Mason package.

I guess my preferred option would be to use the Pkg package with Mason (option 2).
As far I can see, I could probably make an entry in the registry for the missing target and point to a GitHub repo having that release. For the Pkg package, would I need to implement a corresponding provider (PURL) ?
What are your suggestions ?

Ps: I was looking for input prior to writing here i.e. googled, reviewed Mason GitHub issues, checked the Mason plugin help in nvim. Only saw that there is an issue for local repositories i.e.  Add the ability to point to a local repository #1424, which could be one solution strategy.

**Log:**
```
11 [DEBUG Tue Jul 25 22:53:29 2023] ...te/pa
    ck/packer/start/mason.nvim/lua/mason-core
    /fetch.lua:37: Fetching URL "https://api.
    mason-registry.dev/api/github/mason-org/m
    ason-registry/releases/latest"↴
 12 [DEBUG Tue Jul 25 22:53:29 2023] .../pack
    /packer/start/mason.nvim/lua/mason-core/p
    rocess.lua:116: Spawning cmd="curl", spaw
    n_opts={↴
 13   args = { "-H", "Accept: application/vnd
    .mason-registry.v1+json; q=1.0, applicati
    on/json; q=0.8", "-H", "User-Agent: mason
    .nvim v1.6.0 (+https://github.com/william
    boman/mason.nvim)", "-fsSL", "-X", "GET",
     "--connect-timeout", 30, "https://api.ma
    son-registry.dev/api/github/mason-org/mas
    on-registry/releases/latest" }↴
 14 }↴
15 [DEBUG Tue Jul 25 22:53:29 2023] .../pack
    /packer/start/mason.nvim/lua/mason-core/p
    rocess.lua:162: Spawned with pid 8638↴
 16 [DEBUG Tue Jul 25 22:53:31 2023] .../pack
    /packer/start/mason.nvim/lua/mason-core/p
    rocess.lua:148: Job pid=8638 exited with
    exit_code=0, signal=0↴
 17 [DEBUG Tue Jul 25 22:53:31 2023] ...te/pa
    ck/packer/start/mason.nvim/lua/mason-core
    /fetch.lua:37: Fetching URL "https://gith
    ub.com/mason-org/mason-registry/releases/
    download/2023-07-25-fake-coast/registry.j
    son.zip"↴
 18 [DEBUG Tue Jul 25 22:53:31 2023] .../pack
    /packer/start/mason.nvim/lua/mason-core/p
    rocess.lua:116: Spawning cmd="curl", spaw
    n_opts={↴
19   args = { "-H", "User-Agent: mason.nvim
    v1.6.0 (+https://github.com/williamboman/
    mason.nvim)", "-fsSL", "-X", "GET", "-o",
     "/data/data/com.termux/files/home/.local
    /share/nvim/mason/registries/github/mason
    -org/mason-registry/registry.json.zip", "
    --connect-timeout", 30, "https://github.c
    om/mason-org/mason-registry/releases/down
    load/2023-07-25-fake-coast/registry.json.
    zip" }↴
 20 }↴
 21 [DEBUG Tue Jul 25 22:53:31 2023] .../pack
    /packer/start/mason.nvim/lua/mason-core/p
    rocess.lua:162: Spawned with pid 8641↴
 22 [DEBUG Tue Jul 25 22:53:32 2023] .../pack
    /packer/start/mason.nvim/lua/mason-core/p
    rocess.lua:148: Job pid=8641 exited with
    exit_code=0, signal=0↴
24 [DEBUG Tue Jul 25 22:53:32 2023] ...te/pa
    ck/packer/start/mason.nvim/lua/mason-core
    /fetch.lua:37: Fetching URL "https://gith
    ub.com/mason-org/mason-registry/releases/
    download/2023-07-25-fake-coast/checksums.
    txt"↴
 25 [DEBUG Tue Jul 25 22:53:32 2023] .../pack
    /packer/start/mason.nvim/lua/mason-core/p
    rocess.lua:116: Spawning cmd="curl", spaw
    n_opts={↴
 26   args = { "-H", "User-Agent: mason.nvim
    v1.6.0 (+https://github.com/williamboman/
    mason.nvim)", "-fsSL", "-X", "GET", "--co
    nnect-timeout", 30, "https://github.com/m
    ason-org/mason-registry/releases/download
    /2023-07-25-fake-coast/checksums.txt" }↴
 27 }↴
28 [DEBUG Tue Jul 25 22:53:32 2023] .../pack
    /packer/start/mason.nvim/lua/mason-core/p
    rocess.lua:162: Spawned with pid 8646↴
 29 [DEBUG Tue Jul 25 22:53:32 2023] .../pack
    /packer/start/mason.nvim/lua/mason-core/p
    rocess.lua:148: Job pid=8646 exited with
    exit_code=0, signal=0↴
 30 [INFO  Tue Jul 25 22:53:34 2023] ...acker
    /start/mason.nvim/lua/mason-core/installe
    r/init.lua:184: Executing installer for P
    ackage(name=lua-language-server) {}↴
 31 [DEBUG Tue Jul 25 22:53:34 2023] ...acker
    /start/mason.nvim/lua/mason-core/installe
    r/init.lua:58: Attempting to lock package
     Package(name=lua-language-server)↴
 32 [DEBUG Tue Jul 25 22:53:34 2023] ...acker
    /start/mason.nvim/lua/mason-core/installe
    r/init.lua:70: Wrote lockfile Package(nam
    e=lua-language-server)↴
33 [DEBUG Tue Jul 25 22:53:34 2023] .../site
    /pack/packer/start/mason.nvim/lua/mason-c
    ore/fs.lua:71: fs: mkdirp /data/data/com.
    termux/files/home/.local/share/nvim/mason
    /staging/lua-language-server↴
 34 [DEBUG Tue Jul 25 22:53:34 2023] ...rt/ma
    son.nvim/lua/mason-core/installer/registr
    y/init.lua:148: Compiling installer. lua-
    language-server {}↴
 35 [DEBUG Tue Jul 25 22:53:34 2023] ...rt/ma
    son.nvim/lua/mason-core/installer/registr
    y/init.lua:109: Parsing spec lua-language
    -server {}↴
 36 [DEBUG Tue Jul 25 22:53:34 2023] ...rt/ma
    son.nvim/lua/mason-core/installer/registr
    y/init.lua:140: Failed to parse spec spec
     lua-language-server PLATFORM_UNSUPPORTED
37 [DEBUG Tue Jul 25 22:53:34 2023] .../site
    /pack/packer/start/mason.nvim/lua/mason-c
    ore/fs.lua:59: fs: unlink /data/data/com.
    termux/files/home/.local/share/nvim/mason
    /staging/lua-language-server.lock↴
 38 [DEBUG Tue Jul 25 22:53:34 2023] .../site
    /pack/packer/start/mason.nvim/lua/mason-c
    ore/fs.lua:47: fs: rmrf /data/data/com.te
    rmux/files/home/.local/share/nvim/mason/s
    taging/lua-language-server↴
 39 [DEBUG Tue Jul 25 22:53:34 2023] ...acker
    /start/mason.nvim/lua/mason-core/installe
    r/init.lua:39: Building receipt for Packa
    ge(name=lua-language-server)↴
 40 [ERROR Tue Jul 25 22:53:34 2023] ...acker
    /start/mason.nvim/lua/mason-core/installe
    r/init.lua:249: Installation failed for P
    ackage(name=lua-language-server) **error="T
    he current platform is unsupported**."
```↴

**Comment by williamboman** [comment]
Hello! So pre-built assets for Termux are rarely (if ever) provided. Termux provides their own package manager which you can use to download the language server. Note that they need to apply patches to make it build, I'm not sure how invasive these are or if the lua-langauge-server maintainers are interested in providing Termux/Bionic binaries.

**Reply by aBaechtold** [reply]
Thank you William, I agree that it would be better to handle this by the tool (lua_ls) itself, but that the Termux platform is probably not very interesting for the maintainers. As I mentioned, someone already made the efforts to create a Termux package, so that's what I use now. I just wondered whether I could obtain that package from within Mason with moderate efforts.
